File: src/test/java/com/github/albertosh/flitetrakr/sync/usecases/cheapestconnection/CheapestConnectionUseCaseTest.java
package com.github.albertosh.flitetrakr.sync.usecases.cheapestconnection;

import com.github.albertosh.flitetrakr.sync.services.IConnectionService;
import com.github.albertosh.flitetrakr.sync.usecases.MockService;

import org.junit.Before;
import org.junit.Test;

import static org.hamcrest.MatcherAssert.assertThat;
import static org.hamcrest.Matchers.contains;
import static org.hamcrest.Matchers.not;
import static org.hamcrest.Matchers.nullValue;
import static org.hamcrest.core.Is.is;

public class CheapestConnectionUseCaseTest {

    private IConnectionService service;
    private CheapestConnectionUseCase cheapestConnectionUseCase;

    @Before
    public void setUp() throws Exception {
        service = MockService.doMock();
        cheapestConnectionUseCase = new CheapestConnectionUseCase(service);
    }

    @Test
    public void existingConnection() throws Exception, CheapestConnectionUseCaseError {
        // Given

        // When
        CheapestConnectionUseCaseInput input = new CheapestConnectionUseCaseInput.Builder()
                .from("NUE")
                .to("AMS")
                .build();
        CheapestConnectionUseCaseOutput result = cheapestConnectionUseCase.execute(input);

        // Then
        assertThat(result.getCities(), contains("NUE", "FRA", "AMS"));
        assertThat(result.getPrice(), is(60));
    }

    @Test
    public void sameOriginAndDestiny() throws Exception, CheapestConnectionUseCaseError {
        // Given

        // When
        CheapestConnectionUseCaseInput input = new CheapestConnectionUseCaseInput.Builder()
                .from("LHR")
                .to("LHR")
                .build();
        CheapestConnectionUseCaseOutput result = cheapestConnectionUseCase.execute(input);

        // Then
        assertThat(result.getCities(), contains("LHR", "NUE", "FRA", "LHR"));
        assertThat(result.getPrice(), is(93));
    }

    @Test
    public void nonExistingConnection() throws Exception {
        // Given

        // When
        CheapestConnectionUseCaseInput input = new CheapestConnectionUseCaseInput.Builder()
                .from("NUE")
                .to("DXB")
                .build();

        CheapestConnectionUseCaseError error = null;
        try {
            cheapestConnectionUseCase.execute(input);
        } catch (CheapestConnectionUseCaseError e) {
            error = e;
        }
        // Then
        assertThat(error, is(not(nullValue())));
        assertThat(error.getErrorType(),
                is(CheapestConnectionUseCaseError.ErrorType.CONNECTION_NOT_FOUND));
    }

}

File: flitetrakr_sync_kotlin/src/test/kotlin/com/github/albertosh/flitetrakr/usecases/CheapestConnectionUseCaseTest.kt
package com.github.albertosh.flitetrakr.usecases

import com.github.albertosh.flitetrakr.services.IConnectionService
import io.kotlintest.specs.ShouldSpec

class CheapestConnectionUseCaseTest : ShouldSpec() {

    init {
        "CheapestConnectionUseCase" {

            should("throw an error if the connection doesn't exist") {
                val service: IConnectionService = mockService()
                val usecase = CheapestConnectionUseCase(service)

                val exception: CheapestConnectionUseCaseError? =
                        try {
                            usecase.execute(
                                    CheapestConnectionUseCaseInput(
                                            from = "NUE",
                                            to = "DXB"
                                    ))
                            null
                        } catch (e: CheapestConnectionUseCaseError) {
                            e
                        }

                exception shouldBe CheapestConnectionUseCaseError.connectionNotFound
            }

            should("return the minimum price of the connection") {
                val service: IConnectionService = mockService()
                val usecase = CheapestConnectionUseCase(service)

                val output = usecase.execute(
                        CheapestConnectionUseCaseInput(
                                from = "NUE",
                                to = "AMS"
                        ))

                output shouldBe CheapestConnectionUseCaseOutput(
                        cities = listOf("NUE", "FRA", "AMS"),
                        price = 60
                )
            }

            should("look for a connection around the globe if the origin and destination is the same") {
                val service: IConnectionService = mockService()
                val usecase = CheapestConnectionUseCase(service)

                val output = usecase.execute(
                        CheapestConnectionUseCaseInput(
                                from = "LHR",
                                to = "LHR"
                        ))

                output shouldBe CheapestConnectionUseCaseOutput(
                        cities = listOf("LHR", "NUE", "FRA", "LHR"),
                        price = 93
                )
            }

        }
    }
}
